### PR TITLE
ref: Refactor AsyncComponent to TypeScript

### PR DESCRIPTION
**Pull request by lynnagara**
This should make it easier to write components that extend
AsyncComponent. For example https://github.com/getsentry/sentry/pull/14125.

Usage:
```
class Foo extends AsyncComponent<FooProps, FooState & AsyncComponentState> {
  renderBody() {
    // ...
  }
}
```

### CODE REVIEW

File: src/sentry/static/sentry/app/components/asyncComponent.tsx [hunk 1]
```diff
@@ -13,7 +13,25 @@ import PermissionDenied from 'app/views/permissionDenied';
 import RouteError from 'app/views/routeError';
 import getRouteStringFromRoutes from 'app/utils/getRouteStringFromRoutes';
 
-export default class AsyncComponent extends React.Component {
+type Props = {
+  location?: any;
```
Comment (markstory) [inline src/sentry/static/sentry/app/components/asyncComponent.tsx]: Should this use the [Location interface](https://github.com/Microsoft/TypeScript/blob/master/lib/lib.dom.d.ts#L9516)?

File: src/sentry/static/sentry/app/components/asyncComponent.tsx [hunk 2]
```diff
@@ -144,8 +162,11 @@ export default class AsyncComponent extends React.Component {
     document.removeEventListener('visibilitychange', this.visibilityReloader);
   }
 
+  api: any;
```
Comment (markstory) [inline src/sentry/static/sentry/app/components/asyncComponent.tsx]: Should we have an interface for the ApiClient?

**Review by dashed — CHANGES_REQUESTED** [review]

File: src/sentry/static/sentry/app/components/asyncComponent.tsx (hunk 1 above)
Comment (dashed) [reply src/sentry/static/sentry/app/components/asyncComponent.tsx]: Be aware that the `location` object may be sourced from instances of react-router v3.

File: src/sentry/static/sentry/app/components/asyncComponent.tsx (hunk 2 above)
Comment (dashed) [reply src/sentry/static/sentry/app/components/asyncComponent.tsx]: I tried to type the `Client` object during Hack Week, but I wasn't able to get too far. 

A `.d.ts` file may be provided as an alternative.

Comment (lynnagara) [reply src/sentry/static/sentry/app/components/asyncComponent.tsx]: Added this to the discussion items for the next TSC. I agree getting type definitions for this class should be a priority.

**Review by dashed — APPROVED** [review]
Looks good to me 👍

File: src/sentry/static/sentry/app/components/asyncComponent.tsx [hunk 3]
```diff
@@ -13,7 +14,25 @@ import PermissionDenied from 'app/views/permissionDenied';
...
+  errors: object;
+  remainingRequests?: number;
+  [key: string]: any;
+};
+
+export default class AsyncComponent<
+  P extends Props = Props,
+  S extends AsyncComponentState = AsyncComponentState
+> extends React.Component<P, S> {
   static propTypes = {
```
Comment (lynnagara) [inline src/sentry/static/sentry/app/components/asyncComponent.tsx]: Leaving the prop types in here until the usage of this component is fully converted to TS

**Pull request merged**